<a href="https://colab.research.google.com/github/PeggyK1/DS-Unit-1-Sprint-2-Statistics/blob/master/Peggy_Krom_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [0]:
### YOUR WORK HERE

In [22]:
#Data
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data

--2020-06-08 20:32:05--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25936 (25K) [application/x-httpd-php]
Saving to: ‘imports-85.data.1’

imports-85.data.1   100%[===================>]  25.33K  --.-KB/s    in 0.1s    

2020-06-08 20:32:06 (178 KB/s) - ‘imports-85.data.1’ saved [25936/25936]



In [0]:
#Pandas and Numpy
import pandas as pd
import numpy as np

In [24]:
#Setting Column headers and reading in dataset
column_headers = ['symbol', 'norm_loss', 'make', 'fuel_type', 'asp', 'doors', 'bod_style', 'dr_wheels', 'eng_loc', 'wheel_base', 'length', 'width', 'height', 'curb_wt', 'eng_type', 'cyl', 'eng_size', 'fuel_sys', 'bore', 'stroke', 'comp_ratio', 'hp', 'peak_rpm', 'city_mpg', 'hiwy_mpg', 'price']

df = pd.read_csv('imports-85.data', names=column_headers)
df.head()

symbol norm_loss         make fuel_type  ... peak_rpm city_mpg hiwy_mpg  price
0       3         ?  alfa-romero       gas  ...     5000       21       27  13495
1       3         ?  alfa-romero       gas  ...     5000       21       27  16500
2       1         ?  alfa-romero       gas  ...     5000       19       26  16500
3       2       164         audi       gas  ...     5500       24       30  13950
4       2       164         audi       gas  ...     5500       18       22  17450

[5 rows x 26 columns]

In [25]:
#Null values represented as NaN
df = df.replace({'?':np.NaN})
df.head()

symbol norm_loss         make fuel_type  ... peak_rpm city_mpg hiwy_mpg  price
0       3       NaN  alfa-romero       gas  ...     5000       21       27  13495
1       3       NaN  alfa-romero       gas  ...     5000       21       27  16500
2       1       NaN  alfa-romero       gas  ...     5000       19       26  16500
3       2       164         audi       gas  ...     5500       24       30  13950
4       2       164         audi       gas  ...     5500       18       22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [0]:
### YOUR WORK HERE

In [26]:
#Sample of 10 cars
sample = df.sample(10, random_state=30)
sample

symbol norm_loss           make  ... city_mpg hiwy_mpg  price
121       1       154       plymouth  ...       31       38   6692
176      -1        65         toyota  ...       27       32  10898
24        1       148          dodge  ...       31       38   6229
88       -1       137     mitsubishi  ...       23       30   9279
72        3       142  mercedes-benz  ...       16       18  35056
103       0       108         nissan  ...       19       25  13499
26        1       148          dodge  ...       31       38   7609
20        0        81      chevrolet  ...       38       43   6575
155       0        91         toyota  ...       27       32   8778
185       2        94     volkswagen  ...       27       34   8195

[10 rows x 26 columns]

In [27]:
#Mean of Highway-mpg for whole dataset
df['hiwy_mpg'].mean()

30.75121951219512

In [28]:
#Mean of Highway-mpg for sample dataset
sample['hiwy_mpg'].mean()

32.8

In [31]:
sample['hiwy_mpg'].dtypes

dtype('int64')

In [0]:
#Scipystats
from scipy import stats

Null Hypothesis: The mean from the sample is close to the mean of the whole

Alternative Hypothesis: The means are different

Confidence Level: 95%

In [39]:
#Testing sample vs whole
stats.ttest_1samp(sample['hiwy_mpg'], 30.75, nan_policy='omit')

Ttest_1sampResult(statistic=0.8917727404948264, pvalue=0.3957307423595724)

Based on the t-statistic of .89 and the p-value of .4, I ** fail to reject** the null hypothesis because the p-value is greater than 0.05

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [0]:
### YOUR WORK HERE

Null Hypothesis: The average Highway MPG is 35, based on the salesman's claim

Alternative Hypothesis: The average Highway MPG is not 35

Confidence Level: 95%

In [38]:
#Testing Salesman Claim
stats.ttest_1samp(sample['hiwy_mpg'], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

Based on a t-statistic of .95 and a p-value of .36, I **fail to reject** the null hypothesis. I don't have enough information to say that the average highway mpg is not 35.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [0]:
### YOUR WORK HERE

In [35]:
#100 Car Sample
sample2 = df.sample(100, random_state=30)
sample2.head()

symbol norm_loss           make  ... city_mpg hiwy_mpg  price
121       1       154       plymouth  ...       31       38   6692
176      -1        65         toyota  ...       27       32  10898
24        1       148          dodge  ...       31       38   6229
88       -1       137     mitsubishi  ...       23       30   9279
72        3       142  mercedes-benz  ...       16       18  35056

[5 rows x 26 columns]

In [40]:
sample2['hiwy_mpg'].mean()

30.64

In [43]:
stats.ttest_1samp(sample2['hiwy_mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

Based on a t-statistic of 6.51 and a p-value of 3e-09, I **reject* the null hypothesis and suggest the alternative to be true. The average highway MPG is not 35.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

The highway mpg mean for the 10 car sample was 32.8., p-value > 0.05

The highway mpg mean for the 100 car sample was 30.64, p-value < 0.05

The highway mpg mean for the whole car database was 30.75, p-value < 0.05

The 10 car sample was higher leading to the impression that the salesman's claims could be accurate while in truth they were not, compared to the 100 car sample and the whole database.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [0]:
### YOUR WORK HERE

In [44]:
#Loading Data
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-06-08 21:20:31--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-06-08 21:20:32 (126 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [46]:
column_headers2 = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df2 = pd.read_csv('house-votes-84.data', names=column_headers2)

print(df2.shape)
df2.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [47]:
congress = df2.replace({'y': 1, 'n': 0, '?':np.NaN})

congress.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [48]:
congress.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [0]:
#Separating data into two parties
dem = congress[congress['party'] == 'democrat']
rep = congress[congress['party'] == 'republican']

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [0]:
### YOUR WORK HERE

In [54]:
#El Salvador Aid Test
print(stats.ttest_ind(rep['el-salvador-aid'], dem['el-salvador-aid'], nan_policy='omit'))
print(rep['el-salvador-aid'].mean())
print(dem['el-salvador-aid'].mean())

Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68)
0.9515151515151515
0.21568627450980393


In [55]:
#Water Project Test
print(stats.ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit'))
print(rep['water-project'].mean())
print(dem['water-project'].mean())

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)
0.5067567567567568
0.502092050209205


In [56]:
#Budget Test
print(stats.ttest_ind(rep['budget'], dem['budget'], nan_policy='omit'))
print(rep['budget'].mean())
print(dem['budget'].mean())

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)
0.13414634146341464
0.8884615384615384


In [57]:
#Anti Satellite Ban Test
print(stats.ttest_ind(rep['anti-satellite-ban'], dem['anti-satellite-ban'], nan_policy='omit'))
print(rep['anti-satellite-ban'].mean())
print(dem['anti-satellite-ban'].mean())

Ttest_indResult(statistic=-12.526187929077842, pvalue=8.521033017443867e-31)
0.24074074074074073
0.7722007722007722


In [58]:
#Education Test
print(stats.ttest_ind(rep['education'], dem['education'], nan_policy='omit'))
print(rep['education'].mean())
print(dem['education'].mean())

Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64)
0.8709677419354839
0.14457831325301204


1. Budget: p-value=2.07e-77 t-stat=23.21 Dem Support: 89% Rep Support: 13%

p-value: 2.07e-77 < p-value: 0.01 with higher dem mean

2. Education: p-value=1.88e-64 t-stat=20.5 Dem Support: 14% Rep Support: 87%

p-value: 1.88e-64 < p-value: 0.01 with higher rep mean

3. Water Project: p-value=0.93 t-stat=.089 Dem Support: 50% Rep Support: 51%

p-value: 0.93 > p-value: 0.01 Similar Dem and Rep means

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [0]:
### YOUR WORK HERE

In [0]:
#Worked on after review, don't count. I did it wrong the first time
for col in df.columns[1:]:
  output.append(stats.ttest_ind(rep[col], dem[col], nan_policy='omit'))



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [0]:
### YOUR WORK HERE

In [60]:
sample2['hiwy_mpg'].std()

6.688630485964803

In [0]:
import math

In [63]:
t = (sample2['hiwy_mpg'].mean() - 35) / (sample2['hiwy_mpg'].std() /math.sqrt(100))
print(t)

-6.518524246703232


In [64]:
#Testing with Scipy
stats.ttest_1samp(sample2['hiwy_mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [0]:
### YOUR WORK HERE

In [0]:
a = dem['crime'].dropna()

In [0]:
b = rep['crime'].dropna()

In [75]:
a.shape

(257,)

In [76]:
b.shape

(161,)

In [78]:
t = (a.mean() - b.mean()) / (math.sqrt(a.std()/257 + b.std()/161))
print(t)

-12.141711253531504


In [79]:
#Checking with scipy
stats.ttest_ind(a, b)

Ttest_indResult(statistic=-16.342085656197696, pvalue=9.9523427056044e-47)